cuda/base/scoped_device_id.cpp

```diff
@@ -58,14 +58,11 @@ cuda_scoped_device_id_guard::cuda_scoped_device_id_guard(int device_id)
 }
 
 
-cuda_scoped_device_id_guard::~cuda_scoped_device_id_guard() noexcept(false)
+cuda_scoped_device_id_guard::~cuda_scoped_device_id_guard()
 {
     if (need_reset_) {
-        /* Ignore the error during stack unwinding for this call */
-        if (std::uncaught_exception()) {
-            cudaSetDevice(original_device_id_);
-        } else {
-            GKO_ASSERT_NO_CUDA_ERRORS(cudaSetDevice(original_device_id_));
+        if (cudaSetDevice(original_device_id_) != cudaSuccess) {
+            std::terminate();
```

Comment:
any reason?

Reply:
I want to prevent throwing from a destructor here, since it is strongly discouraged and perhaps even UB. So when we are in a case where we can't reset the device id, I can't think of a way to gracefully handle that error. So instead I just terminate the program, which would happen if we throw here anyway.

Reply:
The user might set the device id to something different for ginkgo and other parts of their code. If we can't restore the user's device id, I don't think we should just silently continue. In general, I don't think there is a reasonable way to continue execution, if we can't reset the device id to what the user expects.
This combined with the core guidelines (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-dtor-fail) on this made me chose to terminate here.

Reply:
I think turning this into an exception or failing loudly is still preferable to the silent failure we have at the moment - either `std::exit(1234)` or `throw` or a combination thereof would be preferable IMO. Throwing inside a destructor has somewhat well-defined semantics (stack unwind if in normal code, terminate if in stack unwinding process), so I think we can do the same as we do in `Executor::raw_free`

Reply:
I will do the same as in `Executor::raw_free`, i.e. use `std::exit` with a more explicit error message. This object is designed such that the destructor should always be part of the normal stack unwinding process, throwing an exception here will always lead to `std::terminate`. Also it seems that we would need C++-17 to reliable check if it's safe to throw in the destructor, see https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4152.pdf.